Page List: Add ability to only show child pages

mkaz opened the pull request:
## Description

This scratches my own itch, I have a set of tutorial pages that has a single parent page, and then children pages off that parent. I want to build a table of contents that just include a listing of all the child pages.

Example [in old theme here](https://mkaz.blog/working-with-vim/) if curious the way it was built in the old theme is [in this template page](https://github.com/mkaz/mcbain/blob/trunk/page-tut.php).

So for a block theme, I want to use the Pages List to do the same, but need to be able to filter that block and only show the children of the current page.

Fixes #31063

**Implementation ?** This uses `global $post` to get the `$parent_id` which may not be ideal, but `post-comments` and `latest-posts` both use the same. Open to suggestions on a better way to fetch parent id.

## How has this been tested?

1. Create new page and a child of that page
2. Add Navigation block to the parent page
3. Add Page List to the page
4. Confirm that all pages show (default behavior)
5. Toggle Show children only
6. Confirm that only the child page shows


## Types of changes

- Adds new childrenOnly attribute to PageList block
- Updates server-side render to use attribute and filter pages

mkaz committed:
Update Page List with toggle to show child pages

- Adds a showOnlyChildPages attribute to Edit
- Update render to use attribute and only pull in children
- Do not show on post pages which will not have a parent
- Allow showing on wp_template pages

Fixes #31063

gwwar commented on packages/block-library/src/page-list/index.php:
```diff
@@ -146,27 +146,37 @@ function nest_pages( $current_level, $children ) {
 	static $block_id = 0;
 	$block_id++;
 
+	$children_only = ( isset( $attributes['childrenOnly'] ) && $attributes['childrenOnly'] );
```
Do you mind also adding support for this option in the PageList -> NavigationLinks conversion? This was added in https://github.com/WordPress/gutenberg/pull/30390

mkaz replied:
🤔 Looking at the convert to links, it is only allowed if the parent block is a navigation block - it took me a bit to figure this out because my test case is using the PageList in a widget in a sidebar, I didn't wrap it in a Navigation block.

mkaz replied:
The way the Site Editor work now, I don't think we can do this. When in the editor and you click convert to links it will not have the context of what page you are in - so it will hard code the links in the template, but we want it to use the parent of the page you are browsing.

I created a short video that hopefully explains (with audio)

https://user-images.githubusercontent.com/45363/118560925-1ccd0700-b71f-11eb-8a03-b52043b20e8f.mp4

gwwar commented on packages/block-library/src/page-list/index.php:
```diff
@@ -146,27 +146,37 @@ function nest_pages( $current_level, $children ) {
  * @return string Returns the page list markup.
  */
 function render_block_core_page_list( $attributes, $content, $block ) {
+	global $post;
```
One way of avoiding global, would be storing a `parent` attribute rather than `childrenof` boolean.

We can still show a boolean toggle in the block settings, but we'd internally serialize the parent post id.

This would match the query parameter option of the pages api https://developer.wordpress.org/rest-api/reference/pages/

<img width="770" alt="Screen Shot 2021-05-17 at 8 54 22 AM" src="https://user-images.githubusercontent.com/1270189/118519095-a5cb4a80-b6ed-11eb-9802-64bbec88f18e.png">

mkaz replied:
I don't think this will work either because it will fix the menu to a specific page, the same issue highlighted in the video above. It won't let a single template be used for multiple pages with children pages, which is the use case I'm aiming to solve in this PR.

The following pages:
https://mkaz.blog/working-with-vim
https://mkaz.blog/working-with-rust
https://mkaz.blog/working-with-go

All use the same template and the menu is created from the child pages of the parent.

gwwar commented:
Oh gotcha, this is filtering relative to the current page 👍 

So taking a step back a bit, my question then is PageList a good fit, or would we like to add this elsewhere? It could make sense to add this feature in the block, but we'd also need to test this nested in the navigation block, and in various editors (widget, navigation, site, post ). cc @tellthemachines @jasmussen @talldan @getdave 

Would it more sense to extend capabilities in the Query component and loop to support this? I tested this out with the site editor and there are a few missing pieces to make this possible:
- Add an advanced filter to filter by the children of the current page in Query component. (This might still need to use global $post)
- Maybe a better loop component for these links. Internally these pieces have access to postIds in the loop with `"usesContext": [ "postId" ]`.

See example in: 
https://github.com/WordPress/gutenberg/blob/a8061c1db1e4252fb14d0ee033cf3340d694b114/packages/block-library/src/post-title/block.json#L8

<img width="1319" alt="Screen Shot 2021-05-18 at 12 13 19 PM" src="https://user-images.githubusercontent.com/1270189/118710071-aab5fa00-b7d2-11eb-824c-758add2cd18c.png">

cc @ntsekouras

jasmussen commented:
I can't fully figure out why you'd show only subpages inside the navigation block. But I also can't rule out that you might want to do this in exotic situations, so I'm not fundamentally opposed. But I can't think up how to make an intuitive UI for it; what would happen with a page list inside a navigation block that sits in a header template part but is set to show only subpages? It would likely show nothing, right? For that reason it might be best to not show the option when inside the navigation block.

I want to also bring the attention briefly to #31879, and an adjacent conversation that spawned from the navigation overlay color property, on whether it might make sense to rewrite the Page List as a client block. I'm not making any value judgemnets on which approach is feasible or best here, I'll leave that to you — merely bring it up in case we're optimizing in either direction.

talldan commented:
I created a couple of issues that I thought would be useful for filtering Page List as well (https://github.com/WordPress/gutenberg/issues/29584, https://github.com/WordPress/gutenberg/issues/31063). I agree it all seems a bit like the query block. A suggestion I had in the past was to look into using Query to implement Page List.

> I want to also bring the attention briefly to #31879, and an adjacent conversation that spawned from the navigation overlay color property, on whether it might make sense to rewrite the Page List as a client block. 

As in a static block? The array of blocks can change length, so I think it'd always need a dynamic wrapper block, but I guess the items could be static blocks?

jasmussen commented:
To my understanding: still a dynamic block as the point is to keep it in sync with new or deleted pages on the site. CC @georgeh as he had looked into it for the overlay color.

ntsekouras commented on packages/block-library/src/page-list/edit.js:
```diff
@@ -78,8 +84,23 @@ export default function PageListEdit( { context, clientId } ) {
+				<PanelBody>
+					<ToggleControl
+						label={ __( 'Show only children' ) }
+						checked={ childrenOnly }
```
```suggestion
						checked={ !! childrenOnly }
```

ntsekouras commented:
TBH I got a bit confused with the output in editor vs front-end in my testing, but I have also not experimented much with Navigation block and more 'complex' scenarios like lot's of  parent/children relationships, deeply nested etc...

Having said that an alternative for sure could be extra controls in Query and maybe another one could be having a block variation for `Page List` to show just the children of current page?

mkaz commented:
> TBH I got a bit confused with the output in editor vs front-end in my testing, but I have also not experimented much with Navigation block and more 'complex' scenarios like lot's of parent/children relationships, deeply nested etc...

Agree, I see this in the other template editing with dynamic content tii. I'm not sure if we have a way to set context for it or not.

Overall, I think people will want to use the child page feature to generate links, likewise for the additional filters that @talldan mentions in #29584  I'm not sure if it fits in Query Block or not, quite similar but my mindset was that I wanted a list of pages for navigation, so Page List was what first came to mind.

mkaz commented:
The video below shows some additional controls that will be required — not necessarily in the form you see, but stubbing them out to illustrate. The PageList block can be used in multiple contexts, for example in a Post, on a Page, or in a Widget area.

https://user-images.githubusercontent.com/45363/135008698-96807864-f690-4608-9164-d8c8ed7f41da.mp4

The video above shows an example on a post so the context toggle is disabled by default, when inserted in a Page it could be toggled and disable the Page search control.

The goal of the "children only" toggle is to allow specifying a parent page, and then the Page List will only include the children of the page specified.

The context control is to show the children of the page being viewed without having to specify the page. This is looking forward to block themes that may want to create a single template for pages that show sub menus for children pages.

I welcome thoughts on this approach and suggestions on a UX for setting and explaining. Additionally, keeping in mind filters for categories and tags?

jasmussen commented:
Looks like it could work. 

We need a better label than "Show only children". Nothing great comes to mind, perhaps "Limit to showing subpages"?

The second half, "Use current page as parent" is where much of the complexity comes in. Would it be at all possible to only surface the "show only child pages" of the _current_ page, and then explore separately whether we need additional controls at all?

mkaz commented:
@jasmussen @gwwar I've updated the behavior and I think it is close to where we want it. We may want to edit the wording on the toggle and help, open to suggestions.

The video shows the Page List under three scenarios
1. On a Post - the toggle does not show. :+1: 
2. On a Page - the toggle shows, and updates the list based on child pages. What do we show when no child pages? :question: 
3. On a template - the toggle shows, but has no context on what to show in page :question: 

https://user-images.githubusercontent.com/45363/135177604-666d6458-09d6-4792-8ba7-7599d7836362.mp4

I want to update the unit tests with these scenarios, but technically it should be ready for review?

jasmussen commented:
Nice. I think we can tweak the verbiage in the inspector a bit, and if feasible without too many big changes, can you make the controls a part of the main inspector card? I.e. no separator between description and content? Do this only if it doesn't require hacky CSS:
<img width="364" alt="page list" src="https://user-images.githubusercontent.com/1204802/135233200-0c3ed593-666f-4b2c-a9ba-ec5412bc3dc7.png">

mkaz committed:
Update verbiage explaining control

mkaz committed:
Simplify if-else for setting top/child pages, makes linter happier

mkaz commented:
## 1. About the underline in panel

If I remove the `<PanelBody/>` wrapper around the toggle it removes the line but I also lose all the padding. See:

![image](https://user-images.githubusercontent.com/45363/135303094-213b3a25-0bec-461f-a7a3-e548188dc519.png)

## 2. About verbiage

I updated the text using your verbiage but keeping "child" instead of "sub". I like keeping it because Parent is the name used in the editor, in various functions ([example](https://developer.wordpress.org/reference/functions/get_page_children/)), and in external documentation ([one example](https://wordpress.com/support/pages/page-options/#parent-page))

jasmussen commented:
Cool, to keep things simple let's just keep the panelbody for now and the project will circle back as part of #27331.

gwwar commented:
This looks promising @mkaz I did find a few corner cases while manually testing:

- Toggling List Child Pages when a page has no children will make the block unselectable (since it has no height).

https://user-images.githubusercontent.com/1270189/135317110-98fde7ee-b700-41bb-90cf-39f2ebfeae94.mp4

- Convert logic in Navigation Block doesn't have awareness of this filtering:

https://user-images.githubusercontent.com/1270189/135317470-1067f39b-42b3-456f-abd0-03b2fb765f79.mp4

> On a template - the toggle shows, but has no context on what to show in page ❓

Should we be able to see the option in a template? I didn't spot it while testing.

mkaz commented:
Fixed issue with empty block (no child pages) in https://github.com/WordPress/gutenberg/pull/35441

tellthemachines reviewed (commented):
Thanks for working on this! Testing the changes, there are a couple of unexpected things happening when "Limit to child pages" is set:

* On top-level pages only children of the current page appear, but on child pages we can see both children _and_ siblings, plus the siblings' whole descendant tree. This is confusing, especially because the description states "When enabled, the block lists only child pages of the current page."

* Children are rendered as a flat list, even when there are multiple descendant levels. I'd expect hierarchy to be reflected in the list markup, as it is when Page List renders all pages.

mkaz commented:
Thanks @tellthemachines for reviewing, I'll look at further nesting of pages.

With regards to the first behavior, that is what is desired, so maybe needs better explanatory labels. The idea is to enable creating a menu based on the parent-child relationship, so an entire section can be browsed. For example, if you navigate into one of the children, you should be able to navigate to other siblings. This is particularly useful for documentation sites.

mkaz committed:
Reverse logic on when to display toggle

The toggle was set to show explictly on certain posts types but it is
easier to hide on post types so it will show on others such was
templates, template parts, or menus.

mkaz committed:
Update logic for checking top-level and child pages

Moves check for top level and child checks to its own function so it
came be a little more clear what it is doing and why.

When the only_child_pages flag is set, we need to treat direct
child pages to the parent as top level so they show. And treat grand
children pages as child pages so they nest.

mkaz commented:
I fixed the nesting, so if there are "grand child" pages then they will show nested with the toggle flipped.

I'm open to suggestions on naming the toggle, maybe @jasmussen and "Limit to sub-pages" is more accurate since it's not just children. The description could be: "When enabled, the block will list the sub-pages of the top most parent"  🤷‍♂️

tellthemachines commented:
> With regards to the first behavior, that is what is desired, so maybe needs better explanatory labels. The idea is to enable creating a menu based on the parent-child relationship, so an entire section can be browsed

It feels a bit random though, because siblings at any nesting level may or may not be part of a section. E.g. if I have:

- top level
- - child 1
- - - grandchild 1 1
- - - grandchild 1 2
- - child 2
- - - grandchild 2 1
- - - grandchild 2 2

it's not clear that "child 1" tree is relevant when we're browsing "child 2"; that depends fully on how your site is structured.

We could make "also show sibling pages" an optional toggle when "show only child pages" is enabled. What feels odd is the unexplained discrepancy in behaviour between a top level and a child page. Hierarchies can have many levels, and the point at which it makes sense to show sibling pages will vary from site to site.

> The description could be: "When enabled, the block will list the sub-pages of the top most parent"

This is not quite accurate either. In the example above, if I set a page list to only show child pages in "grandchild 1 1", it'll show only itself and "grandchild 1 2", not any of its ancestors.

jasmussen commented:
Good points all around. Just to zoom out a bit and ponder a different approach, could the toggle be inverted and say "Omit top level pages"? It would start off by default, and we'd need a good paragraph of helptext explaining the use case.

This is also a case where it's okay to be fairly verbose with the description. Site Logo already spends quite a few lines of text on its own:
<img width="306" alt="Screenshot 2021-10-12 at 09 31 14" src="https://user-images.githubusercontent.com/1204802/136911656-71e4b81c-452a-40d1-a231-e2d757408b38.png">

mkaz committed:
Fix rebasing issues

scruffian commented:
This has now been achieved in https://github.com/WordPress/gutenberg/pull/45861
